File: 08/day8.pl
:- use_module(library(clpfd)).

input(Input) :-
    read_file_to_string("day8.input", String, []),
    string_chars(String, Chars),
    select('\n', Chars, CharsNoNewline),
    maplist([In, Out]>>number_chars(Out, [In]), CharsNoNewline, Input).

layers([], [], _).
layers(Input, [Layer|Layers], Size) :-
    length(Layer, Size),
    append(Layer, Rem, Input),
    layers(Rem, Layers, Size).

zeroes_ones_twos(Layer, Zeroes, Ones, Twos) :-
    Cardinalities = [0-Zeroes,1-Ones,2-Twos],
    global_cardinality(Layer, Cardinalities).

part1(Layers, Ans) :-
    maplist([In, Z-O-T]>>zeroes_ones_twos(In, Z, O, T), Layers, Counts),
    keysort(Counts, [Z-O-T|_]),
    Ans #= O * T.

apply_layer([], [], []).
apply_layer([HLayer|TLayer], [HImage|TImage], Out) :-
    apply_layer(TLayer, TImage, Temp),
    ( (var(HImage), HLayer\=2)
    -> 
        Out = [HLayer|Temp]
    ;
        Out = [HImage|Temp]
    ).

part2(Layers) :-
    length(Image, 150),
    foldl(apply_layer, Layers, Image, Out),
    maplist([In, Out]>>(In=1->Out='x';Out=' '), Out, Ascii),
    layers(Ascii, Img, 25),
    maplist([In]>>(string_chars(Str, In), writeln(Str)), Img).

run :-
    input(Input),
    layers(Input, Layers, 150), % = 25 * 6
    part1(Layers, Out1),
    format("Part 1: ~w~n", Out1),
    writeln("Part 2:"),
    part2(Layers).

File: 08/day8.rs
use std::fs::File;
use std::io;
use std::io::Read;

fn parse(filename: &str) -> io::Result<Vec<usize>> {
    let mut file = File::open(filename)?;
    let mut input = String::new();
    file.read_to_string(&mut input)?;
    let v: Vec<usize> = input.trim_end().chars().map(|c| c.to_digit(10).unwrap() as usize).collect();
    Ok(v)
}

fn part1(input: &Vec<usize>) -> usize {
    let mut ans = 0;
    let mut fewest = 150; // = 25 * 6
    for layer in input.chunks(150) {
        let zeroes = layer.iter().filter(|&x| *x == 0).count();
        let ones = layer.iter().filter(|&x| *x == 1).count();
        let twos = layer.iter().filter(|&x| *x == 2).count();
        if zeroes >= fewest {
           continue 
        }
        fewest = zeroes;
        ans = ones * twos;
    }
    ans
}

fn part2(input: &Vec<usize>) {
    let mut image: [usize; 150] = [2; 150];
    for layer in input.chunks(150) {
        for (i, c) in layer.iter().enumerate() {
            if image[i] != 2 {
                continue
            }
            if *c == 2 {
                continue
            }
            image[i] = *c;
        }
    }
    for row in image.chunks(25) {
        let prettyprint: Vec<char> = row.iter().map(|&x| if x==0 { ' ' } else { '\u{2588}' }).collect();
        let s: String = prettyprint.into_iter().collect();
        println!("{}", s);
    }
}

fn main() {
    let input = parse("day8.input").unwrap();
    println!("Part 1: {}", part1(&input));
    println!("Part 2:");
    part2(&input);
}
